pkg/virt-handler/virtwrap/manager.go

```diff
@@ -357,3 +382,101 @@ func (l *LibvirtDomainManager) setDomainXML(vm *v1.VirtualMachine, wantedSpec ap
 	}
 	return dom, nil
 }
+
+// getDomainDesc takes a domain pointer and domain xml flags and returns an api.DomainSpec
+func (l *LibvirtDomainManager) getDomainDesc(dom cli.VirDomain, flags libvirt.DomainXMLFlags) (*api.DomainSpec, error) {
+
+	xmlstr, err := dom.GetXMLDesc(flags)
+	if err != nil {
+		return nil, err
+	}
+
+	var newSpec api.DomainSpec
+	err = xml.Unmarshal([]byte(xmlstr), &newSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &newSpec, nil
+}
+
+// createNetworkingMetadata adds networking metadata to the domain XML.
+func createNetworkingMetadata(domain *api.DomainSpec) error {
+
+	for i, iface := range domain.Devices.Interfaces {
+		if iface.Type == "nodeNetwork" {
+			ifmeta := api.InterfaceMetadata{
+				Type:  "nodeNetwork",
+				Index: i,
+			}
+			domain.Metadata.Interfaces.Interfaces = append(domain.Metadata.Interfaces.Interfaces, ifmeta)
+			// We cant configure the network yet, set something here so that libvirt let us persist the metatada
+			iface.Type = "ethernet"
+			domain.Devices.Interfaces[i] = iface
+		}
+	}
+	return nil
+}
+
+// CNIAdd takes a domain, looks for CNI networks and tries to do a CNI ADD
+func (l *LibvirtDomainManager) CNIAdd(domain *api.DomainSpec) error {
+	for idx, ifmeta := range domain.Metadata.Interfaces.Interfaces {
+		// TODO somehow group network types which require CNI in the API,
+		// so that we don't have to hardcode the config names.
+		if ifmeta.Type == "nodeNetwork" {
+
+			iface := domain.Devices.Interfaces[ifmeta.Index]
+
+			obj, err := model.Clone(&iface)
+			if err != nil {
+				return err
+			}
+
+			newIf := obj.(*api.Interface)
+
+			// Allow overriding the mac address if one is specified.
+			// If not present, leave it to the plugin to select/generate an appropriate mac
+			mac := ""
+			if newIf.MAC != nil {
+				mac = newIf.MAC.MAC
+			}
+
+			result, err := l.cniTool.CNIAdd(fmt.Sprintf("%s_%d", domain.Name, ifmeta.Index), "nodenetwork", "not important", &mac, 1)
+			if err != nil {
+				return fmt.Errorf("error invoking CNI Add: %v", err)
+			}
+
+			// TODO return the device type , so that we don't have to hardcode "direct"
+			newIf.Type = "direct"
+			newIf.Source = api.InterfaceSource{
+				Device: result.Interfaces[0].Name,
+				Mode:   "bridge",
+			}
+			newIf.MAC = &api.MAC{MAC: result.Interfaces[0].Mac}
+			domain.Metadata.Interfaces.Interfaces[idx].IP = result.IPs[0].Address.IP.String()
```

Comment:
@rmohr It took me awhile to understand exactly what is happening here, but I think I get it all now. There are two DHCP IP address requests going on in this current design.

1. The local host's dhcp client daemon which attempts to reserve an IP for the mac address used by a VM.

2. The VM guest itself will request the an IP over DHCP using the same MAC during initialization.

From our discussion earlier, the concern we have is that we aren't 100% guaranteed that the IP address returned and picked by request 1 and 2 will be identical.  The two requests/responses are identical in our vagrant setups, but that's not a guarantee we can make across all networks. 

The solution you and @vladikr are working on would involve a local DHCP server that would respond to guest VM DHCP requests. Essentially, the KubeVirt runtime would be asking the network for an IP address for a specific MAC, then we'd use a local DHCP server to pass that IP to the guest VM in order to ensure the right IP is used for the guest.

Is my understanding correct here?